File: Lab_08/sol-lab08/java/task-2/src/Main.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Scanner;
import java.util.ArrayList;
import java.util.Collections;

public class Main {
	static class Task {
		public static final String INPUT_FILE = "in";
		public static final String OUTPUT_FILE = "out";
		public static final int NMAX = 100005; // 10^5

		int n;
		int m;

		@SuppressWarnings("unchecked")
		ArrayList<Integer> adj[] = new ArrayList[NMAX];

		private void readInput() {
			try {
				Scanner sc = new Scanner(new BufferedReader(new FileReader(
								INPUT_FILE)));
				n = sc.nextInt();
				m = sc.nextInt();

				for (int i = 1; i <= n; i++)
					adj[i] = new ArrayList<>();
				for (int i = 1; i <= m; i++) {
					int x, y;
					x = sc.nextInt();
					y = sc.nextInt();
					adj[x].add(y);
					adj[y].add(x);
				}
				sc.close();
			} catch (IOException e) {
				throw new RuntimeException(e);
			}
		}

		private void writeOutput(ArrayList<Integer> result) {
			try {
				PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter(
								OUTPUT_FILE)));
				for (int node : result) {
					pw.printf("%d ", node);
				}
				pw.printf("\n");
				pw.close();
			} catch (IOException e) {
				throw new RuntimeException(e);
			}
		}

		private int depth[];
		private int low[];

		private void dfs(int x, int f, ArrayList<Integer> sol) {
			depth[x] = depth[f] + 1;
			low[x] = depth[x];
			boolean ok = false;
			// Cnt = numarul de copii al lui x in arborele DFS.
			int cnt = 0;
			for (int y : adj[x]) {
				if (y == f) continue;
				if (depth[y] == 0) {
					dfs(y, x, sol);
					cnt++;
					if (low[y] >= depth[x]) {
						// Acest nod este posibil sa fie punct critic (caz particular
						// - radacina), deoarece nu au existat muchii de intoarcere catre
						// vreun stramos al lui x.
						ok = true;
					}
					// updatez low_link:
					// - low_link[x] = depth-ul cel mai mic pe care X il
					// cunoaste
					// - low_link[y] = depth-ul cel mai mic pe care Y il
					// cunoaste
					// Tot ce acceseaza y (Ex. y - ... - z) poate accesa si x
					// (ex. x - y - ... - z)
					low[x] = Math.min(low[x], low[y]);
				} else {
					// Muchie de intoarcere.
					low[x] = Math.min(low[x], depth[y]);
				}
			}

			// Daca e radacina si are cel putin 2 copii e punct critic
			if (f == 0 && cnt > 1) sol.add(x);
			// Sau daca nu e radacina si ok == true.
			else if (f != 0 && ok) sol.add(x);
		}

		private ArrayList<Integer> getResult() {
			// TODO: Gasiti nodurile critice ale grafului neorientat stocat cu liste
			// de adiacenta in adj.
			ArrayList<Integer> sol = new ArrayList<>();
			depth = new int[n + 1];
			low = new int[n + 1];
			for (int i = 0; i <= n; i++) depth[i] = 0;
			for (int i = 1; i <= n; i++) {
				if (depth[i] != 0) continue;
				dfs(i, 0, sol);
			}
			return sol;
		}

		public void solve() {
			readInput();
			writeOutput(getResult());
		}
	}

	public static void main(String[] args) {
		new Task().solve();
	}
}

File: Lab_08/sol-lab08/cpp/task-2/main.cpp
// Tarjan - O(n + m)

#include <bits/stdc++.h>
using namespace std;

const int kNmax = 100005;
#define edge pair<int, int>
#define MIN(x, y) ((x) < (y) ? (x) : (y))

class Task {
public:
  void solve() {
    read_input();
    print_output(get_result());
  }

private:
  // n = numar de noduri, m = numar de muchii
  int n, m;

  // adj[node] lista de adiacenta a nodului node pentru graful initial
  vector<int> adj[kNmax];

  // bcc[i] = componenta biconexa cu indicele i
  vector<vector<int>> bcc;

  // ordinea de vizitare
  // found[node] = timpul de start a lui node in parcurgerea DFS
  // in laborator found se numeste idx
  vector<int> found;

  // low_link[node] = min { found[x] | x este accesibil din node }
  // adica timpul minim al unui nou
  vector<int> low_link;

  // vector in care retin punctele de articulatie
  vector<int> cut_vertex;
  vector<int> is_cv;

  // parent[i] = parintele nodului i
  vector<int> parent;

  void read_input() {
    ifstream fin("in");
    fin >> n >> m;

    found = vector<int>(n + 1, -1);
    low_link = vector<int>(n + 1, 0);
    parent = vector<int>(n + 1, 0);
    is_cv = vector<int>(n + 1, 0);

    for (int i = 1, x, y; i <= m; i++) {
      fin >> x >> y;
      adj[x].push_back(y);
      adj[y].push_back(x);
    }
    fin.close();
  }

  vector<int> get_result() {
    /*
    TODO: Gasiti nodurile critice ale grafului neorientat stocat cu liste
    de adiacenta in adj.
    */
    tarjan();

    return cut_vertex;
  }

  void tarjan() {
    // momentul curent de start
    // pe masura ce vizita nodurile el va creste (++)
    int current_start = 0;

    for (int i = 1; i <= n; ++i) {
      if (found[i] == -1) {
        // acest nod nu a fost descoperit, deci il putem folosi
        // marcam nodul i ca fiind radacina
        parent[i] = 0;

        // pornim o noua cautare din nodul respectiv
        dfs(i, current_start);
      }
    }
  }

  void dfs(int node, int &current_start) {
    // incep un nou nod, deci un nod timp de start
    ++current_start;  // numarul nodului 

    // atat found, cat si low_link vor primi valoarea lui current_start
    found[node] = current_start;
    low_link[node] = current_start;

    // initializez numarul de copii al nodului curent cu 0
    int children = 0;

    for (auto &vecin : adj[node]) {
      if (found[vecin] == -1) { // deci il pot vizita
        // parintele nodului in care ma duc este chiar nodul curent
        parent[vecin] = node;

        // cresc numarul de copii
        ++children;

        // pornesc un nou dfs
        dfs(vecin, current_start);

        // updatez low_link:
        // - low_link[node]  = timpul de start cel mai mic pe care NODE  il
        // cunoaste
        // - low_link[vecin] = timpul de start cel mai mic pe care VECIN il
        // cunoaste
        // Tot ce acceseaza vecin (Ex. vecin - ... - x) poate accesa si node
        // (ex. node - vecin - ... - x)
        low_link[node] = MIN(low_link[node], low_link[vecin]);

        // CUT VERTEX - caz 2:
        // daca nu au existat muchii de intoarcere spre un STRAMOS al lui node
        // inseamna ca nodul curent este punct de articulatie
        // adica VECIN nu a putut sa sara inapoi in arbore PESTE node
        if (low_link[vecin] >= found[node]) {
          // radacina este tratata separat (adica conditia de mai sus nu este
          // pentru radacina)
          if (parent[node] != 0) {
            // deci node este CV
            // il daugat la solutie doar daca nu a fost deja adaugat
            if (!is_cv[node]) {
              cut_vertex.push_back(node);
              is_cv[node] = 1;
            }
          }
        }
      } else {
        if (vecin != parent[node]) {
          // am gasit o muchie de intoarcere
          low_link[node] = MIN(low_link[node], found[vecin]);
        }
      }
    }

    // CUT VERTEX - caz 1:
    // daca nodul curent este radacina si are mai mult de 2 copii
    if (parent[node] == 0 && children >= 2) {
      // salvez pe node doar daca nu l-am salvat deja
      if (!is_cv[node]) {
        is_cv[node] = 1;
        cut_vertex.push_back((node));
      }
    }
  }

  void print_output(vector<int> result) {
    ofstream fout("out");
    for (int i = 0; i < int(result.size()); i++) {
      fout << result[i] << ' ';
    }
    fout << '\n';
    fout.close();
  }
};

int main() {
  Task *task = new Task();
  task->solve();
  delete task;
  return 0;
}
